ffi/ffi_wrapper.go

```diff
@@ -17,35 +18,267 @@ import (
+	if err != nil {
+		return C.CString(fmt.Sprintf("FFI lib error: parsing resolver address failed: %+v", err))
+	}
+	resolver := register.NewDefaultRestResolverClient(addr)
+
+	agentDid := C.GoString(cAgentDid)
+	agentKeyName := C.GoString(cAgentKeyName)
+	agentName := C.GoString(cAgentName)
+	agentSeed := C.GoString(cAgentSeed)
+
+	agentIdentity, _, err = getIdentity(api.GetAgentIdentity, agentDid, agentKeyName, agentName, agentSeed)
+	if err != nil {
+		return C.CString(fmt.Sprintf("FFI lib error: failed to get agent registered identity: %+v", err))
+	}
+
+	userDid := C.GoString(cUserDid)
+	userKeyName := C.GoString(cUserKeyName)
+	userName := C.GoString(cUserName)
+	userSeed := C.GoString(cUserSeed)
+
+	userIdentity, _, err = getIdentity(api.GetUserIdentity, userDid, userKeyName, userName, userSeed)
+	if err != nil {
+		return C.CString(fmt.Sprintf("FFI lib error: failed to get agent registered identity: %+v", err))
+	}
+
+	err = api.UserDelegatesAuthenticationToAgent(resolver, userIdentity, agentIdentity, delegationName)
+
+	if err != nil {
+		return C.CString(fmt.Sprintf("FFI lib error: unable to delegate control to agent: %+v", err))
+	}
+	return nil
+}
+
+//export TwinDelegatesControlToAgent
+func TwinDelegatesControlToAgent(cResolverAddress *C.char,
+	cAgentDid *C.char,
+	cAgentKeyName *C.char,
+	cAgentName *C.char,
+	cAgentSeed *C.char,
+
+	cTwinDid *C.char,
+	cTwinKeyName *C.char,
+	cTwinName *C.char,
+	cTwinSeed *C.char,
+
+	cDelegationName *C.char) *C.char {
+
+	resolverAddress := C.GoString(cResolverAddress)
+	delegationName := C.GoString(cDelegationName)
+
+	var twinIdentity register.RegisteredIdentity
+	var agentIdentity register.RegisteredIdentity
+
+	addr, err := url.Parse(resolverAddress)
+	if err != nil {
+		return C.CString(fmt.Sprintf("FFI lib error: parsing resolver address failed: %+v", err))
+	}
+	resolver := register.NewDefaultRestResolverClient(addr)
+
+	agentDid := C.GoString(cAgentDid)
+	agentKeyName := C.GoString(cAgentKeyName)
+	agentName := C.GoString(cAgentName)
+	agentSeed := C.GoString(cAgentSeed)
+
+	agentIdentity, _, err = getIdentity(api.GetAgentIdentity, agentDid, agentKeyName, agentName, agentSeed)
+	if err != nil {
+		return C.CString(fmt.Sprintf("FFI lib error: failed to get agent registered identity: %+v", err))
+	}
+
+	twinDid := C.GoString(cTwinDid)
+	twinKeyName := C.GoString(cTwinKeyName)
+	twinName := C.GoString(cTwinName)
+	twinSeed := C.GoString(cTwinSeed)
+
+	twinIdentity, _, err = getIdentity(api.GetTwinIdentity, twinDid, twinKeyName, twinName, twinSeed)
+	if err != nil {
+		return C.CString(fmt.Sprintf("FFI lib error: failed to get agent registered identity: %+v", err))
+	}
+
+	err = api.TwinDelegatesControlToAgent(resolver, twinIdentity, agentIdentity, delegationName)
+
+	if err != nil {
+		return C.CString(fmt.Sprintf("FFI lib error: unable to delegate control to agent: %+v", err))
+	}
+	return nil
+}
+
+//export IsAllowedFor
+func IsAllowedFor(
+	cResolverAddress *C.char,
+	cToken *C.char,
```

Comment:
(consistency) For another function above, the last argument was on the same line as the close brace, here not. (I prefer close brace on next line - but maybe gofmt disagrees?)

Reply:
done